### ISSUE TITLE
Interface overhaul for a breaking 2.0 release

### ISSUE THREAD

**Issue by HDembinski** [scikit-hep/iminuit#220]
If we make a 2.0 release, we should use the opportunity to completely break the interface and make it more consistent and Pythonic. There are several issue with the current interface.

* legacy attributes with ambiguous names like `args` and `matrix`
* multiple ways of accessing the same thing, e.g. `args`, `values`; `matrix`, `covariance`
* exposed internal attributes and methods
  - `frontend`
  - `refreshInternalState()`
  - `initialParameterState()`, also see #163
  - `construct_FCN()`
  - `cfmin`
  - `last_upst`
  - `minimizer`
* duplication of functionality
  - PythonFCN counts function calls, Minuit internally also counts function calls
  - `pos2var` and `var2pos` duplicate internal C++ conversion utils
* `errordef` should only be settable in `__init__`, and only two values should be allowed `1` for least-squares and `0.5` for likelihoods
* Replace setters and getters with properties (setters and getters are not pythonic, see e.g. https://python-course.eu/python3_properties.php or talks by Python experts like Raymond Hettinger)
  - `(set|get)_strategy()`
  - `(set|get)_print_level()`
  - ...
* `merrors` format is awkward, it should just return a tuple for parameter with the error_minus, error_plus, see #153

Also see #105

**Comment by cdeil** [comment]
@HDembinski - Thanks for bringing this up!

Apparently when @piti118 wrote iminuit, backward compatibility with pyminuit was a goal (see e.g. comment [here](https://github.com/iminuit/iminuit/blob/aa628c4ace0f17cd6249aab0e4e1be0e54014277/iminuit/_libiminuit.pyx#L98) and in other places in the code or docstrings), and he favoured that over a cleaner API.

I think we can all agree that there are some things that can be improved if breaking backward compatibility is on the table. A lot of good suggestions by @HDembinski above. But before we start discussing those, let's discuss the bigger question: should we ever do a backward-incompatible change in iminuit? Or should we treat is as a legacy codebase / API, just like probably Minuit2 isn't changed even if they know of things that could be improved in the API now.

Clearly there are pros and cons, i.e. existing users hate if their scripts break (or scripts / pipeline code they inherited from someone that has left and that has been running OK for years), and new users that learn `iminuit` get confused e.g. by duplicate attributes or other usability issues with the current API.

There might be a middle ground, with the 1.x release as-is, just bug fixes, and a 2.x release with improvements such as the ones listed above. But of course, that can also be a dis-service to users, like the Python 2 -> 3 split that we had for the past 10 years. Another option could be that @HDembinski does a new project, because I think really you would prefer the use of pybind11 over Cython anyways, right? But there it's even more important: is the better, clean API worth the split?

I have no good feeling on this. I'm inclined to suggest to keep iminuit 100% backward compatible, and to mainly improve the docs. But I'm not sure if this is the best course of action, I agree that a simpler, cleaner iminuit 2.0 is very appealing, especially if @HDembinski or someone else is actually offering to actually do the work. (I have no time, except for small edits or to review PRs that come in and give some feedback.)

I'm not aware of any big projects that rely on iminuit that should get a say in this decision. I think it's mostly individual researchers or small groups and it's very hard to get opinions what our users want here?

@piti118 @jpivarski and all - please give your opinion here!

**Comment by jpivarski** [comment]
Coincidentally, I informally asked a group of about 10 physicists actively working on CMS analyses how many have ever accessed Minuit directly in its capacity as a function minimizer, rather than indirectly through fitting applications. The answer was 1.5 (one had caveats: he was _fixing_ a broken application).

I'm therefore inclined to think that LHC-style analyses, at least, are now at the level of sophistication that the user interface is at the layer of tools like RooFit, HiggsCombiner, and now TensorFlow, rather than function fitting directly. By contrast, my thesis (which motivated the existence and API of PyMinuit) was a single lineshape fit and it was therefore very important for me to control how data points (distinct beam runs) were handled.

This situation, where a physicist needs to directly control the function minimization might still be common in medium energy physics, where unique conditions can be arranged without moving a thousand-member collaboration, or in other fields, but in LHC-style HEP, it's mostly buried under applications that would be upset by an API change. Those application developers would try to stay on the old branch.

If it's a thin interface change, like variable names or easily translatable behaviors, you may be able to maintain both in one package, like (but not exactly like!) the way ROOT maintains both 6-character Fortran-style names and long SEAL-C++ style names in the TMinuit class. More Pythonically,

```python
import iminuit.v1 as iminuit
```

for the old and

```python
import iminuit
```

for the new. Then in response to an application developer who wants to you too maintain a v1 branch indefinitely (which is extra work for you), you respond with the new import line (which is not). The contents of the `iminuit.v1` submodule would be thin wrappers to the modern API, which can be pretty easy to mock up in Python, owing to Python's dynamism.

I don't know exactly what API changes you have in mind— likely good ones, as I wasn't a good API designer 10 or more years ago when I wrote PyMinuit— do those API changes look like user inputs could be easily translated from the old API to the new with a couple of classes in `iminuit.v1`?

**Comment by jpivarski** [comment]
Oh! All I had to do was scroll up!

Most of these suggestions are definitely good things. Eliminating access to internal state is a good thing, though it would complicate my `iminuit.v1` proposal, since the `v1` module would somehow have to access this state to continue to provide it. Maybe make `v1` incomplete and see who complains, to find out just how many holes you need to poke?

My only suggestion would be that you leave it possible to set `errordef` to values other than 0.5 and 1, though these are clearly special values that should be set by name. The whole reason a user is using Minuit directly (through iminuit, rather than indirectly through RooFit) is to control such things. Maybe they're hand-coding their own regularization, like lasso or ridge regression but for nonlinear fits instead of linear ones, and they need to modify this parameter to keep it statistically meaningful in light of the fact that they're putting extra constraints on the loss function. Things like that.

**Comment by cdeil** [comment]
@jpivarski - Thanks for chiming in.

Yes, just to clarify:  in `iminuit`, we only use / support the bundled `Minuit2` (updated from ROOT a week ago). We do not support the old SEAL Minuit or linking against ROOT directly. So on the MINUIT side, there is no question for us what to do.

The question here is only about the `iminuit` API, and the suggestions are to improve and clean it up like @HDembinski suggested above in the original issue description.

One more thing worth considering is IMO a new API to interact with Minuit in a more generic way (See e.g. https://github.com/iminuit/iminuit/pull/190) that makes it easier to use iminuit / MINUIT e.g. as one fitting backend. An example would be https://github.com/sherpa/sherpa/pull/194/files#diff-41ea9d6a33909297914686689ad93138R892 where Sherpa is a general modeling / fitting backend, and iminuit / MINUIT would be one optimiser / error estimator to add, but doing this currently is pretty cumbersome.

@jpivarski - not everyone is using RooFit, e.g. I'm an astronomer and we want to use iminuit for parameter and error estimation, but via our own modeling / fitting packages, or our own user scripts where we write our own likelihood functions. So iminuit is being used both directly and via other codes (like Gammapy or Sherpa) that call it in the background.

I'm not sure if the `iminuit.v1` idea is a good idea or even feasible. If I understand right, it doesn't allow for any clean-up in the iminuit codebase, which I think @HDembinski has as one motivation. A separate long-term support 1.x git branch on the other hand seems maintainable to me, especially given that iminuit has been doing pretty good for a few years now with very little edits. There could be releases and people could use `iminuit=1.x` in their version specs to make sure they are getting that version and not the 2.x that would break their scripts. It's hard to say if the maintenance effort for such a 1.x branch is very small (if no bugs are found, we don't have to do anything) or if it's real extra work.

**Comment by cdeil** [comment]
> My only suggestion would be that you leave it possible to set errordef to values other than 0.5 and 1, though these are clearly special values that should be set by name.

If we're discussing the points above, I also think this one needs discussion. One point to keep in mind is that currently users have to set different "errordef" values e.g. if they want to get at certain results e.g. from the methods that compute contours. @HDembinski - If a redesign is on the table, it might make sense to only use errordef on init and offer separate "up" or "sigma" options to all methods that need to have that specified, like the contour computation.

If this is done, to separate the "up" in init from the "up" in certain method calls like merrors or contours, then even dropping the `errordef` option in `init` might be worth discussing. Other packages (e.g. "emcee") don't offer that freedom and Fred James in http://lmu.web.psi.ch/docu/manuals/software_manuals/minuit2/mnerror.pdf wrote this, also suggesting that it's reasonable to ask people to put the factor 2 in their function:

> For likelihood fits, the value up=0.5 corresponds to one–standard–deviation errors. Or, alternatively, F may be defined as −2 log(likelihood), in which case differences in F have the same meaning as for χ2 and up=1.0 is appropriate. The two different ways of introducing the factor of 2 are quite equivalent in MINUIT , and although most people seem to use up=0.5, it is perhaps more logical to put the factor 2 directly into FCN.

In astronomy, sometimes we do put the factor 2 in the likelihood function, see e.g. the "cash statistic" at http://cxc.harvard.edu/sherpa/statistics/ which is how we call the Poisson likelihood. So I'm -1 on named arguments, e.g. calling 0.5 the "likelihood" case. Suggest to either leave as-is (complete freedom, document what to put for the common cases), or to avoid the option to get something simpler (at the cost of larger backward incompatibility).

**Comment by cdeil** [comment]
@piti118 - before we discuss more details, we absolutely need your thoughts here, whether we should drop the discussion and keep iminuit as-is, or pursue these large changes.

**Comment by jpivarski** [comment]
Okay, great! Good luck with the update!

(I was only using the Fortran vs C++ names in TMinuit as an analogy.)

((Also, I was _disappointed_ by what I heard about LHC physicists not using function minimizers directly anymore.))

**Comment by HDembinski** [comment]
Wow, so many comments, I am happy that this thread invited so much response!

@jpivarski Thank you for chiming in. :) I think the current design of iminuit is not bad. The tool as it is, is very successful, so the current iminuit does a good job already. I think, however, that it could be made easier and more Pythonic, which would make it easier to learn, and it would prevent some mistakes without writing extensive documentation. A good design is self-explanatory, think of the iPhone, everyone can use it without reading a fat manual. That is the general idea of good design.

The breaking interface changes I have in mind are along the lines of Scott Meyers "Make an interface easy to use correctly, and hard to use incorrectly", and Python's "There should be only one obvious way to do it" (sorry, slightly paraphrased). 

For what I have in mind, I think I don't need to drop the current interface. I also do not want to start a new project. I think the things I have in mind could be achieved mostly by a wrapper class around the original `Minuit` class. Naturally, if this is going to happen, I would volunteer to do most of the actual the coding. These interface issues I raised are superficial, so the new interface could just call the old one underneath. What I would do in a 2.0 release is more along the lines of `from iminuit import Minuit2`. We would keep the original `Minuit` class around, but change the documentation to only refer to `Minuit2` and not to `Minuit`. The latter would get a docstring saying that it is deprecated and not to be used anymore. This would break no ones' code. This approach would still create some overhead, because we would keep tests for the old interface and have new tests for the new interface. This is always the caveat when you want to keep backward compatibility.

The question is, of course, whether all this is worth it. I personally use iminuit all the time, because I am a statistical expert and power-user. High-level tools like Roofit do not give me the level of control I need to implement my algorithms. I generally do statistics starting from first principles with very successful results. Whenever I tried to use packaged algorithms, I always ran into problems sooner or later because the tool was not designed (or only poorly design) to do what I want, or it was simply wrong. Packaged algorithms like Roofit force you into a cage of their ideas, and often it is poorly documented what they are actually doing in detail. Anyway, I do see the appeal for certain rigid applications and for people who are not statistics experts who know what they are doing.
 
As a power user, I know other power users. Many of them like low-level tools like iminuit, for the same reasons. I cannot give you numbers though. The user base is certainly not that large. I am also wondering who is still using iminuit, since development mostly stopped in the last 2 years. Maybe we could make a survey of some sort?

**Comment by cdeil** [comment]
> We would keep the original Minuit class around, but change the documentation to only refer to Minuit2 and not to Minuit.
> The question is, of course, whether all this is worth it.

I've thought about it, and I'm -1 on introducing a Minuit2 class.

> The breaking interface changes I have in mind are along the lines of Scott Meyers "Make an interface easy to use correctly, and hard to use incorrectly", and Python's "There should be only one obvious way to do it"

For me, the main point why we shouldn't introduce Minuit2 is exactly your second point. You'd introduce a second way to do it. If you think users would all migrate to Minuit2 within a year or two, I think that is unrealistic.

My suggestion would be to little code / API improvements (like the numpy methods you just added) that are backwards incompatible. And the docs could use a lot of love. And a few page paper would be nice. If this really happens, a huge improvement was made. Only then I would consider this huge change proposed here.

If there are things in the improvement list above that can be done in a backwards compat way, or by only breaking something that probably only very few or no people rely on, then +1.

> Maybe we could make a survey of some sort?

I don't think we should. It'll be a lot of work and few responses, you won't reach many users anyways.
Big projects like Numpy or Jupyter or Astropy sometimes try every few years, but for iminuit i would not do it.

**Comment by HDembinski** [comment]
Smooth interface transition is ongoing
